Getting "GenerateDepsFile" task failed unexpectedly with KeyNotFoundException when publishing project

I am getting `"GenerateDepsFile" task failed unexpectedly` with `KeyNotFoundException` when publishing my projects.

Below are the conditions that trigger the issue:
- .NET Core 3.0 Preview 4
- WindowsAzure.Storage 9.3.3 is in a .NET Standard library

This can be fixed by either going back to Preview 3 or add WindowsAzure.Storage directly into the .NET Core project.

Error in build log:
```
C:\Program Files\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): Error MSB4018: The "GenerateDepsFile" task failed unexpectedly.
System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Microsoft.NET.Build.Tasks.DependencyContextBuilder2.CalculateExcludedLibraries() in /_/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs:line 769
   at Microsoft.NET.Build.Tasks.DependencyContextBuilder2.Build() in /_/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs:line 215
   at Microsoft.NET.Build.Tasks.GenerateDepsFile.WriteDepsFileNew(String depsFilePath) in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:line 248
   at Microsoft.NET.Build.Tasks.GenerateDepsFile.ExecuteCore() in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:line 284
   at Microsoft.NET.Build.Tasks.TaskBase.Execute() in /_/src/Tasks/Common/TaskBase.cs:line 35
   at Microsoft.NET.Build.Tasks.GenerateDepsFile.Execute() in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:line 262
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

Sample repository: https://github.com/nguyenquyhy/IssueWindowsAzureStoragePublish

Same error on Azure Pipelines, but with 'NETStandard.Library':
```
##[section]Starting: dotnet publish
==============================================================================
Task         : .NET Core
Description  : Build, test, package, or publish a dotnet application, or run a custom dotnet command. For package commands, supports NuGet.org and authenticated feeds like Package Management and MyGet.
Version      : 2.150.1
Author       : Microsoft Corporation
Help         : [More Information](https://go.microsoft.com/fwlink/?linkid=832194)
==============================================================================
[command]C:\windows\system32\chcp.com 65001
Active code page: 65001
[command]C:\hostedtoolcache\windows\dotnet\dotnet.exe publish d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj -c Release -r linux-x64 --self-contained
Microsoft (R) Build Engine version 16.1.54-preview+gd004974104 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 37.56 ms for d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj.
  Restore completed in 316.6 ms for d:\a\1\s\Libs\DpLib\DpLib.csproj.
  Restore completed in 269.87 ms for d:\a\1\s\Libs\TechWiz.Lib\TechWiz.Lib.csproj.
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(151,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(151,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [d:\a\1\s\Libs\TechWiz.Lib\TechWiz.Lib.csproj]
  DpLib -> d:\a\1\s\Libs\DpLib\bin\Release\netstandard2.0\DpLib.dll
  TechWiz.Lib -> d:\a\1\s\Libs\TechWiz.Lib\bin\Release\netstandard2.0\TechWiz.Lib.dll
  TechWiz.Server -> d:\a\1\s\Apps\TechWiz.Server\bin\Release\netcoreapp3.0\linux-x64\TechWiz.Server.dll
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018: The "GenerateDepsFile" task failed unexpectedly. [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018: System.Collections.Generic.KeyNotFoundException: The given key 'NETStandard.Library' was not present in the dictionary. [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at System.Collections.Generic.Dictionary`2.get_Item(TKey key) [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.NET.Build.Tasks.DependencyContextBuilder2.CalculateExcludedLibraries() in /_/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs:line 769 [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.NET.Build.Tasks.DependencyContextBuilder2.Build() in /_/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs:line 217 [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.WriteDepsFileNew(String depsFilePath) in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:line 248 [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.ExecuteCore() in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:line 284 [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() in /_/src/Tasks/Common/TaskBase.cs:line 35 [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateDepsFile.Execute() in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:line 262 [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
C:\hostedtoolcache\windows\dotnet\sdk\3.0.100-preview4-011223\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(801,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj]
##[error]Error: The process 'C:\hostedtoolcache\windows\dotnet\dotnet.exe' failed with exit code 1
##[error]Dotnet command failed with non-zero exit code on the following projects : d:\a\1\s\Apps\TechWiz.Server\TechWiz.Server.csproj
##[section]Finishing: dotnet publish```

--- a/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs
@@ -769,9 +769,9 @@ namespace Microsoft.NET.Build.Tasks
                         //  There may not be a library in the assets file if a referenced project has
                         //  PrivateAssets="all" for a package reference, and there is a package in the graph
                         //  that depends on the same packge.
-                        if (_dependencyLibraries.ContainsKey(dependencyName))
+                        if (_dependencyLibraries.TryGetValue(dependencyName, out var dependencyLibrary))
                         {
-                            includedDependencies.Add(dependencyName, _dependencyLibraries[dependencyName]);
+                            includedDependencies.Add(dependencyName, dependencyLibrary);
                             foreach (var newDependency in _libraryDependencies[dependencyName])
                             {
                                 dependenciesToWalk.Push(newDependency.Name);

--- a/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder2.cs
@@ -766,10 +766,16 @@ namespace Microsoft.NET.Build.Tasks
                     var dependencyName = dependenciesToWalk.Pop();
                     if (!includedDependencies.ContainsKey(dependencyName))
                     {
-                        includedDependencies.Add(dependencyName, _dependencyLibraries[dependencyName]);
-                        foreach (var newDependency in _libraryDependencies[dependencyName])
+                        //  There may not be a library in the assets file if a referenced project has
+                        //  PrivateAssets="all" for a package reference, and there is a package in the graph
+                        //  that depends on the same packge.
+                        if (_dependencyLibraries.ContainsKey(dependencyName))
                         {
-                            dependenciesToWalk.Push(newDependency.Name);
+                            includedDependencies.Add(dependencyName, _dependencyLibraries[dependencyName]);
+                            foreach (var newDependency in _libraryDependencies[dependencyName])
+                            {
+                                dependenciesToWalk.Push(newDependency.Name);
+                            }
                         }
                     }
                 }

--- a/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToExcludeAPackageFromPublish.cs
+++ b/src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToExcludeAPackageFromPublish.cs
@@ -12,6 +12,7 @@ using Xunit;
 using System.Xml.Linq;
 using Xunit.Abstractions;
 using System.Collections.Generic;
+using Microsoft.NET.TestFramework.ProjectConstruction;
 
 namespace Microsoft.NET.Publish.Tests
 {
@@ -165,6 +166,38 @@ namespace Microsoft.NET.Publish.Tests
             }
 
             publishDirectory.Should().OnlyHaveFiles(expectedFiles);
+        }
+
+        [Fact]
+        public void TransitiveNetStandardPackageReferenceAndPublishFalse()
+        {
+            var testLibraryProject = new TestProject()
+            {
+                Name = "TestLibrary",
+                IsSdkProject = true,
+                TargetFrameworks = "netstandard2.0"
+            };
+
+            testLibraryProject.PackageReferences.Add(new TestPackageReference("WindowsAzure.Storage", "9.3.3"));
+
+            var testProject = new TestProject()
+            {
+                Name = "TestApp",
+                IsSdkProject = true,
+                IsExe = true,
+                TargetFrameworks = "netcoreapp3.0"
+            };
+
+            testProject.PackageReferences.Add(new TestPackageReference("Newtonsoft.Json", "12.0.2", privateAssets: "all"));
+
+            testProject.ReferencedProjects.Add(testLibraryProject);
+
+            var testAsset = _testAssetsManager.CreateTestProject(testProject)
+                .Restore(Log, testProject.Name);
+
+            var publishCommand = new PublishCommand(Log, Path.Combine(testAsset.TestRoot, testProject.Name));
+
+            publishCommand.Execute().Should().Pass();
         }
     }
 }

